codelists/tests/test_views.py

```diff
@@ -18,18 +20,97 @@ def logged_in_client(client, django_user_model):
+        "reference-TOTAL_FORMS": "1",
+        "reference-INITIAL_FORMS": "0",
+        "reference-MIN_NUM_FORMS": "0",
+        "reference-MAX_NUM_FORMS": "1000",
+        "reference-0-text": "foo",
+        "reference-0-url": "http://example.com",
+        "signoff-TOTAL_FORMS": "1",
+        "signoff-INITIAL_FORMS": "0",
+        "signoff-MIN_NUM_FORMS": "0",
+        "signoff-MAX_NUM_FORMS": "1000",
+        "signoff-0-user": signoff_user.username,
+        "signoff-0-date": "2020-01-23",
+    }
+
+    request = rf.post("/", data=data)
+    request.user = UserFactory()
+    response = CreateCodelist.as_view()(request, project_slug=project.slug)
+
+    assert response.status_code == 302
+    assert response.url == f"/codelist/{project.slug}/test-codelist/"
+
+    assert project.codelists.count() == 1
+    codelist = project.codelists.first()
+    assert codelist.name == "Test Codelist"
+
+    # we should have one reference to example.com
+    assert codelist.references.count() == 1
+    ref = codelist.references.first()
+    assert ref.url == "http://example.com"
+
+    # we should have one signoff by signoff user
+    assert codelist.signoffs.count() == 1
+    signoff = codelist.signoffs.first()
+    assert signoff.user == signoff_user
+
+
+def test_createcodelist_invalid_post(rf):
+    project = ProjectFactory()
+    signoff_user = UserFactory()
+
+    assert project.codelists.count() == 0
+
     csv_data = "code,description\n1067731000000107,Injury whilst swimming (disorder)"
+
+    # missing signoff-0-date
     data = {
         "name": "Test Codelist",
         "coding_system_id": "snomedct",
         "description": "This is a test",
         "methodology": "This is how we did it",
         "csv_data": _build_file_for_upload(csv_data),
+        "reference-TOTAL_FORMS": "1",
+        "reference-INITIAL_FORMS": "0",
+        "reference-MIN_NUM_FORMS": "0",
+        "reference-MAX_NUM_FORMS": "1000",
+        "reference-0-text": "foo",
+        "reference-0-url": "http://example.com",
+        "signoff-TOTAL_FORMS": "1",
+        "signoff-INITIAL_FORMS": "0",
+        "signoff-MIN_NUM_FORMS": "0",
+        "signoff-MAX_NUM_FORMS": "1000",
+        "signoff-0-user": signoff_user.username,
     }
-    rsp = logged_in_client.post(f"/codelist/{p.slug}/", data, follow=True)
-    assertRedirects(rsp, f"/codelist/{p.slug}/test-codelist/2020-07-23-draft/")
+
+    request = rf.post("/", data=data)
+
+    view = CreateCodelist()
+    view.setup(request)
+    with patch("codelists.views.CreateCodelist.all_valid") as mock_all_valid, patch(
+        "codelists.views.CreateCodelist.some_invalid"
+    ) as mock_some_invalid:
+        view.post(request)
+
+    # was the error handler, some_invalid, called?
+    # we're not doing any custom error handling in the form or formsets so no
+    # need to test what Django is already testing
+    mock_all_valid.assert_not_called()
+    mock_some_invalid.assert_called_once()
```

Comment:
I'm not keen on this style of testing -- it couples the tests to the implementation (making implementation changes harder) and doesn't test what the user experiences (so that it's less useful as a specification of behaviour).  If the form is invalid, does the user get a message?  If so, I think that'd be a better thing to test.

Reply:
That's a fair point, I think I'll remove the mocking and not check that outcome other than the response code.  I like your point about "less useful as a specification of behaviour"!  Since Django (and to some degree Crispy) handle rendering error messages for us I don't think we need to check that?

Reply:
> Since Django (and to some degree Crispy) handle rendering error messages for us I don't think we need to check that?

No need to check that Django's doing the right thing, but it's often worth checking that we're using Django correctly.  In this case, checking for the error message is a good way of validating the behaviour of the view.

Let's not hold this PR up any longer though.  I may do a drive-by refactor of the test in the future, or you can if you like.

Reply:
Fixed in 8a48704

Note that I tested the vieww context contained the expected formset instance with errors rather than looking at the HTML output since Crisipy should handle that.